### PR TITLE
Implement config reloading on SIGHUP

### PR DESCRIPTION
**Pull request by fabxc**
`master` is already merged in `fabxc/servdisc`.

Prometheus now does start with a missing config file but logs errors. As an empty config file is valid, should we just allow starting with `-config.file=""` without any complaints? Would be convenient for testing.

### CODE REVIEW

**Comment by matthiasr** [comment]
Should `reloadConfig` return if err != nil instead of trying to load the bad configuration?

**Comment by fabxc** [comment]
Preferably, yes :)

File: main.go [hunk 1]
```diff
@@ -229,9 +212,33 @@ func NewPrometheus() *prometheus {
...
+	glog.Infof("Loading configuration file %s", *configFile)
+
+	conf, err := config.LoadFromFile(*configFile)
+	if err != nil {
+		glog.Errorf("Couldn't load configuration (-config.file=%s): %v", *configFile, err)
+		glog.Errorf("Note: The configuration format has changed with version 0.14, please check the documentation.")
+		return
+	}
+
+	p.webService.StatusHandler.Config = conf.String()
```
Comment (juliusv) [inline main.go]: Doesn't this need to be locked/protected somehow against concurrent `GET`s of `/status`?

File: main.go [hunk 2]
```diff
@@ -252,6 +259,14 @@ func (p *prometheus) Serve() {
...
 
+	hup := make(chan os.Signal)
+	signal.Notify(hup, syscall.SIGHUP)
+	go func() {
+		for range hup {
+			p.reloadConfig()
+		}
+	}()
+
 	notifier := make(chan os.Signal)
```
Comment (juliusv) [inline main.go]: If the above is called `hup` now, can we call this `term` or something like that?

**Comment by juliusv** [comment]
Looks good in general. I think the ideal end state would be to either apply the entire new config or nothing (if there's an error) in some "atomic" way, but this is great as a first step.

**Comment by fabxc** [comment]
Turns out the target manager can always reload without errors.

Maybe we should consider config and rule files as different things in this regard. Best we can do is logging that there's a bad rule file, anyway. But should the config really depend on it?

File: main.go [hunk 3]
```diff
@@ -229,9 +212,30 @@ func NewPrometheus() *prometheus {
...
+
+	conf, err := config.LoadFromFile(*configFile)
+	if err != nil {
+		glog.Errorf("Couldn't load configuration (-config.file=%s): %v", *configFile, err)
+		glog.Errorf("Note: The configuration format has changed with version 0.14, please check the documentation.")
+		return
+	}
+
+	p.webService.StatusHandler.Lock()
+	p.webService.StatusHandler.Config = conf.String()
```
Comment (juliusv) [inline main.go]: How about giving the `StatusHandler` an `ApplyConfig` method? That way you don't have to expose its lock to the outside world...

**Comment by juliusv** [comment]
:+1:

**Comment by fabxc** [comment]
its happening

**Pull request merged**